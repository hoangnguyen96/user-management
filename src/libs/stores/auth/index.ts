import { createWithEqualityFn } from "zustand/traditional";
import { shallow } from "zustand/shallow";
import { persist } from "zustand/middleware";

// Constants
import { LOCAL_STORAGE } from "@app/constants";

// Models
import { AuthUser } from "@app/models";

interface AuthState {
  isAuthenticated: boolean;
  isAdmin: boolean;
  user: AuthUser;
}

interface AuthStore extends AuthState {
  setAuthenticated: (isAuthenticated: boolean) => void;
  setIsAdmin: (isAdmin: boolean) => void;
  setUser: (user: AuthUser) => void;
  clearAuth: () => void;
}

const INITIAL_AUTH_STATE: AuthState = {
  isAuthenticated: false,
  isAdmin: false,
  user: {
    id: "",
    fullName: "",
    company: "",
    token: "",
  },
};

export const useAuthStore = createWithEqualityFn<AuthStore>()(
  persist(
    (set) => ({
      ...INITIAL_AUTH_STATE,
      setAuthenticated: (isAuthenticated) => {
        set({ isAuthenticated });
      },

      setIsAdmin: (isAdmin) => {
        set({ isAdmin });
      },

      setUser: (user) => {
        set({ user });
      },

      clearAuth: () => {
        set({
          ...INITIAL_AUTH_STATE,
        });
      },
    }),
    { name: LOCAL_STORAGE.AUTH }
  ),
  shallow
);
