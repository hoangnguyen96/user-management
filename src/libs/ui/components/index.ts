// Common
export * from "./Common";

// Components
export { default as HeadingContent } from "./HeadingContent";
export { default as LazyFallbackContent } from "./LazyFallbackContent";
export { default as Logo } from "./Logo";
export { default as Footer } from "./Footer";
export { default as FormController } from "./FormController";
export { default as ListCustomersContent } from "./ListCustomersContent";
export { default as ListProductContent } from "./ListProductContent";
export { default as NavigationBar } from "./NavigationBar";
export { default as Temporary } from "./Temporary";
export { default as TitleContent } from "./TitleContent";
export { default as MenuBar } from "./MenuBar";
export { default as ErrorBoundary } from "./ErrorBoundary";
export { default as ModalCreateUserSuccess } from "./Modal/ModalCreateUserSuccess";
