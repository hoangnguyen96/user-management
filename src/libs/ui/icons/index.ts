export * from "./HelpIcon";
export * from "./IncomeIcon";
export * from "./KeyIcon";
export * from "./Members";
export * from "./ProductIcon";
export * from "./PromoteIcon";
export * from "./Screen";
export * from "./SearchIcon";
export * from "./TotalCustomers";
export * from "./UserIcon";
