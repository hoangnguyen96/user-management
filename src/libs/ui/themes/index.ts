import { createTheme } from "@mui/material/styles";
import { breakpointsBase } from "./breakpoints";
import { colors } from "./colors";
import { typographyBase } from "./typography";

export const themeDefault = createTheme({
  breakpoints: breakpointsBase,
  palette: colors,
  typography: typographyBase,

  components: {
    MuiButton: {
      styleOverrides: {
        root: {
          textTransform: "capitalize",
        },
      },
      variants: [
        {
          props: { variant: "contained" },
          style: {
            color: "White",
          },
        },
        {
          props: { variant: "activate" },
          style: {
            backgroundColor: "rgba(22, 192, 152, 0.38)",
            color: "#008767",
            border: "1px solid #00b087",
            "&:hover": {
              backgroundColor: "rgba(22, 192, 152, 0.62)",
            },
          },
        },
        {
          props: { variant: "inactivate" },
          style: {
            backgroundColor: "#ffc5c5",
            color: "#df0404",
            border: "1px solid #df0404",
            "&:hover": {
              backgroundColor: "rgb(253 170 170)",
            },
          },
        },
        {
          props: { variant: "normal" },
          style: {
            backgroundColor: " rgba(245, 245, 245, 1)",
            color: "rgba(64, 75, 82, 1)",
            border: "1px solid #eeeeee",
            "&:hover": {
              backgroundColor: "rgb(219 219 219)",
            },
          },
        },
      ],
      defaultProps: {
        size: "medium",
      },
    },

    MuiTextField: {
      styleOverrides: {
        root: ({ theme }) => ({
          "& .MuiOutlinedInput-root": {
            borderRadius: "12px",
            fontSize: "14px",
            backgroundColor: theme.palette.common.white,
            "& .css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input": {
              padding: "12px",
            },
            "& fieldset": {
              borderColor: "#b5b7C0",
            },
            "&:hover fieldset": {
              borderColor: theme.palette.primary.main,
            },
            "&.Mui-focused fieldset": {
              borderColor: theme.palette.primary.main,
            },
          },
        }),
      },
    },

    MuiSelect: {
      styleOverrides: {
        root: {
          width: "auto",
          flex: 1,
        },
      },
    },

    MuiFormControl: {
      styleOverrides: {
        root: {
          flexDirection: "row",
          justifyContent: "center",
          alignItems: "center",
          backgroundColor: "#f9fbff",
          borderRadius: 12,
          minWidth: 154,
          "& .css-wama2x-MuiFormLabel-root-MuiInputLabel-root": {
            position: "unset",
            transform: "unset",
            left: "unset",
            top: "unset",
          },
          "& .css-cb5d54-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root":
            {
              "& fieldset": {
                border: 0,
              },
              borderColor: "transparent",
              borderWidth: 0,
            },
          "& .css-wxw8j3-MuiFormLabel-root-MuiInputLabel-root": {
            position: "unset",
            transform: "unset",
            left: "unset",
            top: "unset",
            flex: 0.8,
            padding: 10,
            paddingRight: 5,
            fontSize: "12px",
            textAlign: "right",
          },
          "& .css-zbhh1g-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input.MuiSelect-select":
            {
              padding: "11px 32px 11px 0",
              fontSize: "12px",
              fontWeight: 500,
            },
          "& .css-cb5d54-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root.Mui-focused .MuiOutlinedInput-notchedOutline":
            {
              borderColor: "transparent",
              borderWidth: 0,
            },
          "& .css-10d36jd-MuiSvgIcon-root-MuiSelect-icon, & .css-5ajf1d-MuiSvgIcon-root-MuiSelect-icon":
            {
              width: 20,
              height: 20,
            },
          "& .css-1j5jes1-MuiFormLabel-root.Mui-focused": {
            color: "rgba(0, 0, 0, 0.6)",
          },
        },
      },
    },

    MuiPaper: {
      styleOverrides: {
        root: {
          "& .css-ddqxy4-MuiButtonBase-root-MuiMenuItem-root": {
            fontSize: "12px",
          },
        },
      },
    },

    MuiListItemButton: {
      styleOverrides: {
        root: ({ theme }) => ({
          borderRadius: theme.spacing(1),
          "&.Mui-selected": {
            backgroundColor: theme.palette.primary.main,
            color: theme.palette.common.white,
          },
          "&.Mui-selected:hover": {
            backgroundColor: theme.palette.primary.main,
            color: theme.palette.common.white,
          },
        }),
      },
    },

    MuiLink: {
      styleOverrides: {
        root: ({ theme }) => ({
          backgroundColor: "transparent",
          fontSize: "14px",
          textDecoration: "none",
          color: theme.palette.primary.contrastText,
          "&:hover": {
            cursor: "pointer",
          },
        }),
      },
    },

    MuiPagination: {
      styleOverrides: {
        root: ({ theme }) => ({
          "& .MuiPagination-ul": {
            "& .MuiButtonBase-root.MuiPaginationItem-root": {
              fontSize: "12px",
              fontWeight: 500,
              backgroundColor: "#f5f5f5",
              minWidth: "25px",
              height: "24px",
              margin: "0 5px",
              "&.Mui-selected": {
                backgroundColor: theme.palette.primary.main,
                color: "white",
              },
              "& .MuiSvgIcon-root.MuiPaginationItem-icon": {
                fontSize: "12px",
              },
            },
          },
        }),
      },
    },

    MuiContainer: {
      styleOverrides: {
        root: {
          maxWidth: "100%",
          paddingLeft: 0,
          paddingRight: 0,
          "@media (min-width:600px)": {
            paddingLeft: 0,
            paddingRight: 0,
          },
          "@media (min-width:960px)": {
            paddingLeft: 0,
            paddingRight: 0,
          },
          "@media (min-width:1280px)": {
            paddingLeft: 0,
            paddingRight: 0,
          },
          "@media (min-width:1440px)": {
            maxWidth: "1440px",
          },
        },
      },
    },
  },
});
