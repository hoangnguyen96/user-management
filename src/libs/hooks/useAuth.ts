import { useQueryClient } from "@tanstack/react-query";

// Stores
import { useAuthStore } from "@app/stores";

// Api
import { useAuthLogout } from "@app/api";

export const useAuth = () => {
  const queryClient = useQueryClient();

  // Auth
  const clearAuth = useAuthStore((state) => state.clearAuth);

  // Logout
  const { mutate: logOut } = useAuthLogout(clearAuth);

  const handleLogout = () => {
    logOut();
    queryClient.clear();
  };

  return { handleLogout };
};
