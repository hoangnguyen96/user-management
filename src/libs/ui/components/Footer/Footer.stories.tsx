import { Avatar, Box, Link, Stack, SvgIcon, Typography } from "@mui/material";
import KeyboardArrowDownIcon from "@mui/icons-material/KeyboardArrowDown";
import { Meta, StoryObj } from "@storybook/react";
import { avatar1 } from "@app/ui/images";

const Footer = () => (
  <Box width="306px">
    <Box
      sx={{
        marginBottom: "45px",
        padding: "26px 24px 22px",
        borderRadius: "20px",
        backgroundImage: "linear-gradient(to right, #eaabf0 , #5932ea)",
      }}
    >
      <Typography
        fontSize="14px"
        fontWeight={600}
        color="white"
        textAlign="center"
        marginBottom="20px"
      >
        Upgrade to PRO to get access all Features!
      </Typography>
      <Link
        href="#"
        sx={{
          display: "block",
          padding: "10px 0",
          textAlign: "center",
          fontWeight: 600,
          borderRadius: "20px",
          backgroundColor: "white",
          color: "#5932ea",
        }}
      >
        Get Pro Now!
      </Link>
    </Box>
    <Stack
      flexDirection="row"
      justifyContent="flex-start"
      gap="12px"
      alignItems="center"
      sx={{ cursor: "pointer" }}
    >
      <Avatar
        src={avatar1}
        alt="user-management"
        style={{
          width: "42px",
          height: "42px",
          borderRadius: "50%",
        }}
      />

      <Stack flex={1}>
        <Typography fontSize="14px" fontWeight={400}>
          Evano
        </Typography>
        <Typography fontSize="12px" fontWeight={400} color="#757575">
          Project Manager
        </Typography>
      </Stack>

      <SvgIcon component={KeyboardArrowDownIcon} sx={{ color: "#757575" }} />
    </Stack>
  </Box>
);

const meta: Meta<typeof Footer> = {
  component: Footer,
};

export default meta;
type Story = StoryObj<typeof Footer>;

export const FooterBase: Story = {
  args: {},
};
