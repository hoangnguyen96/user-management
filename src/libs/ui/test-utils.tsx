import { PropsWithChildren, ReactElement, ReactNode } from "react";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import { BrowserRouter } from "react-router-dom";
import { RenderOptions, render } from "@testing-library/react";

// ThemeProvider
import { ThemeProvider } from "@mui/material";

// Config
import { themeDefault } from "./themes";

const AllTheProviders = ({ children }: PropsWithChildren) => {
  return <ThemeProvider theme={themeDefault()}>{children}</ThemeProvider>;
};

export const wrapper = ({ children }: { children: ReactNode }) => {
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
      },
    },
  });

  return (
    <ThemeProvider theme={themeDefault()}>
      <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
    </ThemeProvider>
  );
};

const customRender = (ui: ReactElement, options?: RenderOptions) =>
  render(ui, { wrapper: AllTheProviders, ...options });

const customRenderWithQueryClient = (
  ui: ReactElement,
  options?: RenderOptions
) => render(ui, { wrapper, ...options });

export const WrapperRouter = ({ children }: { children: ReactNode }) => (
  <BrowserRouter>{children}</BrowserRouter>
);

export { customRender as render };
export { customRenderWithQueryClient as renderWithQueryClient };
