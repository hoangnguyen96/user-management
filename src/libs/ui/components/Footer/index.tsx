import { MouseEvent, memo, useCallback, useState } from "react";
import {
  Avatar,
  Box,
  Link,
  ListItemIcon,
  Menu,
  MenuItem,
  Stack,
  SvgIcon,
  Typography,
  useMediaQuery,
} from "@mui/material";

// Stores
import { useAuthStore } from "@app/stores";

// Hooks
import { useAuth } from "@app/hooks";

// Components
import KeyboardArrowDownIcon from "@mui/icons-material/KeyboardArrowDown";
import { Logout } from "@mui/icons-material";
import { themeDefault } from "../../themes";
import { avatar1 } from "@app/ui/images";

const Footer = () => {
  const isLargeScreen = useMediaQuery(themeDefault.breakpoints.up("lg"));
  const [anchorEl, setAnchorEl] = useState<null | HTMLElement>(null);
  const open = Boolean(anchorEl);

  // Auth
  const user = useAuthStore((state) => state.user);
  const { handleLogout } = useAuth();

  const handleClick = useCallback((event: MouseEvent<HTMLElement>) => {
    setAnchorEl(event.currentTarget);
  }, []);

  const handleClose = useCallback(() => {
    setAnchorEl(null);
  }, []);

  const handleLogoutApp = useCallback(() => {
    handleLogout();
    handleClose();
  }, []);

  return (
    <Box>
      {isLargeScreen && (
        <Box
          sx={{
            marginBottom: "45px",
            padding: "26px 24px 22px",
            borderRadius: "20px",
            backgroundImage: "linear-gradient(to right, #eaabf0 , #5932ea)",
          }}
        >
          <Typography
            fontSize="14px"
            fontWeight={600}
            color="white"
            textAlign="center"
            marginBottom="20px"
          >
            Upgrade to PRO to get access all Features!
          </Typography>
          <Link
            href="#"
            sx={{
              display: "block",
              padding: "10px 0",
              textAlign: "center",
              fontWeight: 600,
              borderRadius: "20px",
              backgroundColor: "white",
              color: "#5932ea",
            }}
          >
            Get Pro Now!
          </Link>
        </Box>
      )}
      <Stack
        flexDirection="row"
        justifyContent="flex-start"
        gap="12px"
        alignItems="center"
        sx={{ cursor: "pointer" }}
        onClick={handleClick}
      >
        <Avatar
          src={avatar1}
          alt="user-management"
          sx={{
            width: "42px",
            height: "42px",
            borderRadius: "50%",
          }}
        />
        {isLargeScreen && (
          <>
            <Stack flex={1}>
              <Typography fontSize="14px" fontWeight={400}>
                {user.fullName}
              </Typography>
              <Typography fontSize="12px" fontWeight={400} color="#757575">
                {user.company}
              </Typography>
            </Stack>
            <SvgIcon
              component={KeyboardArrowDownIcon}
              sx={{ color: "#757575" }}
            />
          </>
        )}
      </Stack>
      <Menu
        anchorEl={anchorEl}
        id="account-menu"
        open={open}
        onClose={handleClose}
        onClick={handleClose}
        PaperProps={{
          elevation: 0,
          sx: {
            overflow: "visible",
            filter: "drop-shadow(0px 2px 8px rgba(0,0,0,0.32))",
            mt: 1.5,
            width: isLargeScreen ? "245px" : "50px",
            "& .MuiAvatar-root": {
              width: 32,
              height: 32,
              ml: -0.5,
              mr: 1,
            },
          },
        }}
        transformOrigin={{ horizontal: "right", vertical: "top" }}
        anchorOrigin={{ horizontal: "right", vertical: "bottom" }}
      >
        <MenuItem onClick={handleLogoutApp}>
          <ListItemIcon>
            <Logout fontSize="small" />
          </ListItemIcon>
          {isLargeScreen ? (
            <Typography variant="caption">Logout</Typography>
          ) : (
            <></>
          )}
        </MenuItem>
      </Menu>
    </Box>
  );
};
export default memo(Footer);
