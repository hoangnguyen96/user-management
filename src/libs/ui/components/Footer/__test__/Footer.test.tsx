import { renderWithQueryClient } from "@app/ui/test-utils";
import { fireEvent } from "@testing-library/react";

// Stores
import * as stores from "@app/stores";

// Hooks
import * as hooks from "@app/hooks";

// Component
import Footer from "../index";

const mockUser = {
  fullName: "John Doe",
  company: "Example Inc.",
};

jest.mock("@app/stores", () => ({
  useAuthStore: jest.fn(),
}));

jest.mock("@app/hooks", () => ({
  useAuth: jest.fn(),
}));

jest.mock("@mui/material/useMediaQuery", () => jest.fn());

describe("Footer component", () => {
  const useMediaQuery = require("@mui/material/useMediaQuery");

  beforeEach(() => {
    (hooks.useAuth as unknown as jest.Mock).mockReturnValue({
      handleLogout: jest.fn(),
    });
    (stores.useAuthStore as unknown as jest.Mock).mockReturnValue({
      user: mockUser,
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("Should render snapshot correctly", () => {
    expect(renderWithQueryClient(<Footer />)).toMatchSnapshot();
  });

  it("Should does not render upgrade message on small screens", () => {
    useMediaQuery.mockImplementation(() => true);
    const { queryByText } = renderWithQueryClient(<Footer />);

    expect(
      queryByText(/Upgrade to PRO to get access all Features!/i)
    ).toBeInTheDocument();
    expect(queryByText(/Get Pro Now!/i)).toBeInTheDocument();
  });

  it("Should handle opens and closes menu", () => {
    useMediaQuery.mockImplementation(() => true);
    const { getByTestId, getByText } = renderWithQueryClient(<Footer />);

    fireEvent.click(getByTestId("logout-account"));

    expect(getByText("Logout")).toBeInTheDocument();

    fireEvent.click(getByText("Logout"));
  });

  it("Should calls handleLogoutApp on logout", () => {
    useMediaQuery.mockImplementation(() => true);
    const { getByRole, getByText } = renderWithQueryClient(<Footer />);

    fireEvent.click(getByRole("img"));

    expect(getByText(/Logout/i)).toBeInTheDocument();

    fireEvent.click(getByText(/Logout/i));

    const handleLogout = hooks.useAuth().handleLogout;
    expect(handleLogout).toHaveBeenCalled();
  });
});
