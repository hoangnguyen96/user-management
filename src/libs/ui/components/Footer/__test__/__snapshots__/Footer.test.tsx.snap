// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Footer component Should render snapshot correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiBox-root css-0"
      >
        <div
          class="MuiStack-root css-rtrdhm-MuiStack-root"
          data-testid="logout-account"
        >
          <div
            class="MuiAvatar-root MuiAvatar-circular css-251k5v-MuiAvatar-root"
          >
            <img
              alt="user-management"
              class="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
              src="/src/libs/ui/images/avatar1.png"
            />
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiBox-root css-0"
    >
      <div
        class="MuiStack-root css-rtrdhm-MuiStack-root"
        data-testid="logout-account"
      >
        <div
          class="MuiAvatar-root MuiAvatar-circular css-251k5v-MuiAvatar-root"
        >
          <img
            alt="user-management"
            class="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
            src="/src/libs/ui/images/avatar1.png"
          />
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
